Allow custom actionbar colors for each workspace

agrogers opened the pull request:
This adds two new options to the Color settings page that allows the user to specifiy a color of the workspace header in both day and night modes. 

There are some things I don't  know how to do so if someone wants to give me hand that would be great:
- I need to hide the workspace color option when displaying a the color settings for a window. I believe i know where to do this but I don't know how to access the setting to hide it.
- Only the main bible window view changes the title bar color. The other views keep the original black color. These views seems to use the actionbar while the main bible view uses a toolbar. I think it would be good to have the same color header for all views that are related to a workspace (some app settings are global for example)

And there is some install weirdness:
- It runs fine in the emulator
- When i install directly through Android Studio onto my POCO it works fine
- When i create an APK and try and install on my POCO MIUI device it starts but crashes. 
- When i install onto an Android 8 tablet it installs fine but i get a white actionbar (which seems like the same problem others have been getting)

![image](https://user-images.githubusercontent.com/13920678/147560538-31ebcd83-cc93-4e98-ae7d-23c8d65d71fd.png)

![workspace_color](https://user-images.githubusercontent.com/13920678/147560498-6aabfaa8-0dc4-4c3b-a18a-a150639efff4.jpg)

agrogers committed:
Initial attempt  to set actionbar color

agrogers committed:
Removed unneeded code and restored button colors

agrogers commented:
This will need to be meshed with PR #1980 so the status bar color changes as required as well.

agrogers commented:
And if it is included for release then a database migration addition will be required.

agrogers committed:
Update DatabaseContainer.kt

agrogers committed:
Use the workspace color setting not the window setting.

tuomas2 commented:
Interesting! Maybe coloring tint in workspace selector too?

agrogers committed:
Merge branch 'develop' into Modified-workspace-color

tuomas2 committed:
Merge branch 'develop' into Modified-workspace-color

# Conflicts:
#	app/src/main/java/net/bible/android/view/activity/page/MainBibleActivity.kt

tuomas2 committed:
Fix db version

tuomas2 committed:
Revert changes to 57.json

tuomas2 commented:
Tested this now a bit. 

- [x] Wouldn't it be nice to have some indicator color also on Select workspace view?
- [x] Bug (?): New workspaces should have default colors, not current color
- [ ] Default colors should match with current style
- [ ] ~This should not be stored in TextDisplaySettings but rather in WorkspaceSettings. This is not something to be applied to windows, but only to workspaces. Now the color selector is displayed for window settings too.~ maybe we can live with this too. Will think again.
- [ ] Apply color to system top bar too

agrogers commented:
> Interesting! Maybe coloring tint in workspace selector too?

Definitely. And I might even know how to do that now. I will have a look at it sometime soonish.

I have other things I want to color too - not sure you will like that so much though 🙄

tuomas2 reviewed (changes requested):
Needs to be improved before merging. Left tasks in another comment.

tuomas2 commented:
Pushed some commits in this branch, please pull before continuing working on this :-)

tuomas2 commented:
Changed this PR back to draft status now. 

Please change to "ready for review" status when it's ready for review.

agrogers commented:
@tuomas2 , I think that works nicely.
![image](https://user-images.githubusercontent.com/13920678/149610039-5d72c491-6667-4a16-9ef5-9763b0495463.png)

agrogers committed:
Update WorkspaceEntities.kt

I don't know how to import the colors values into this file.

agrogers committed:
Update DatabaseContainer.kt

I agree that technically this is a workspace setting. But .... read comment.

agrogers committed:
Change the color of the workspace list icons

This seems to work nicely for day and night modes.

agrogers commented:
It seems that when Settings are changed via the workspace list those changes are not saved unless you manually push the Save button or click on a workspace. If you push the back button the changes are lost without notification. I don't know how to hook into  the back button.

agrogers commented:
@tuomas2, I have done the best i can on this now. Will need assistance to take it much further. Comments in the code and I sent you an email as well. Thanks

tuomas2 committed:
Merge branch 'develop' into Modified-workspace-color

tuomas2 committed:
Hide workspace color setting from window settings

tuomas2 commented on app/src/main/java/net/bible/service/db/DatabaseContainer.kt:
```diff
@@ -981,7 +981,21 @@ private val MIGRATION_56_57_breaklines_in_notes = object : Migration(56, 57) {
+    //TODO: ??? Techincally the workspace colors should not be in 'text_display_settings'. But I fear moving it out will make complicate things a bit.
+    //      At present all the code that manages the saving and copying and initialising of colors also works well for the new workspace colors
+    //      But if I move these settings out somewhere else i am not sure if i will then have to write code to manage that aspect a well. If we do,
+    //      I am not sure that the gain in 'purity' is worth the extra work and potential maitenance. Plus i am not sure how to do it :)
```
> I am not sure that the gain in 'purity' is worth the extra work and potential maitenance.

I am feeling that you migth be right here.

tuomas2 commented on app/src/main/java/net/bible/service/db/DatabaseContainer.kt:
```diff
@@ -981,7 +981,21 @@ private val MIGRATION_56_57_breaklines_in_notes = object : Migration(56, 57) {
+            val colDefs = "`text_display_settings_colors_dayWorkspaceColor` INTEGER DEFAULT NULL, `text_display_settings_colors_nightWorkspaceColor` INTEGER DEFAULT NULL".split(",")
+            colDefs.forEach {
+                execSQL("ALTER TABLE `Workspace` ADD COLUMN $it")
+                execSQL("ALTER TABLE `PageManager` ADD COLUMN $it")
```
What if PageManager does not have these fields? Will things break too much?

tuomas2 replied:
Hmm it looks like they need to be there for room to work..

agrogers commented:
I never tried without them but I would assume it would create troubles.

On Thu, 20 Jan 2022, 3:43 am Tuomas Airaksinen, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In app/src/main/java/net/bible/service/db/DatabaseContainer.kt
> <https://github.com/AndBible/and-bible/pull/2007#discussion_r787994161>:
>
> > @@ -981,7 +981,21 @@ private val MIGRATION_56_57_breaklines_in_notes = object : Migration(56, 57) {
>          }
>      }
>  }
> -
> +private val MIGRATION_57_58_workspace_colors = object : Migration(57, 58) {
> +    //TODO: ??? Techincally the workspace colors should not be in 'text_display_settings'. But I fear moving it out will make complicate things a bit.
> +    //      At present all the code that manages the saving and copying and initialising of colors also works well for the new workspace colors
> +    //      But if I move these settings out somewhere else i am not sure if i will then have to write code to manage that aspect a well. If we do,
> +    //      I am not sure that the gain in 'purity' is worth the extra work and potential maitenance. Plus i am not sure how to do it :)
> +    override fun doMigrate(db: SupportSQLiteDatabase) {
> +        db.apply {
> +            val colDefs = "`text_display_settings_colors_dayWorkspaceColor` INTEGER DEFAULT NULL, `text_display_settings_colors_nightWorkspaceColor` INTEGER DEFAULT NULL".split(",")
> +            colDefs.forEach {
> +                execSQL("ALTER TABLE `Workspace` ADD COLUMN $it")
> +                execSQL("ALTER TABLE `PageManager` ADD COLUMN $it")
>
> Hmm it looks like they need to be there for room to work..
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AndBible/and-bible/pull/2007#discussion_r787994161>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKGTJUV3OZL6K4HSJY3EZDUW3Z2LANCNFSM5K362MFQ>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

agrogers committed:
Make status bar same color as toolbar

It almost works. The color does not change until another bible window is closed or shown. I don't know how to fix that.

agrogers commented:
I have now got the Status Bar to have the same color as the toolbar. However there is a bug that I cant fix although I think it is an easy one for you. The status bar color does not update immediately when changing workspaces via the menu. It seems to be setting the statusbar color for the menu activity and not the bible window activity. It works fine if i swipe between workspaces.

As a side note, it doesnt look like it works on my Mi phone because I had 'Hide Notch' enabled. This forces the status bar to be black.

![image](https://user-images.githubusercontent.com/13920678/173193055-05ab8e0e-50b5-487d-8fdc-c24bad81de86.png)
![image](https://user-images.githubusercontent.com/13920678/173193084-81dac6ad-ef3a-4b01-a3f6-d272f5eb32b0.png)

Commit:
Merge branch 'develop' into Modified-workspace-color

# Conflicts:
#	app/src/main/java/net/bible/android/view/activity/page/MainBibleActivity.kt
#	app/src/main/java/net/bible/service/db/DatabaseContainer.kt
#	db/schemas/net.bible.android.database.AppDatabase/58.json

Commit:
Fixes and cleanups

tuomas2 commented:
@agrogers changed storing workspaceColor to WorkspaceSettings after all.

agrogers commented:
You are a purist 😉

On Thu, 24 Nov 2022, 7:52 pm Tuomas Airaksinen, ***@***.***>
wrote:

> @agrogers <https://github.com/agrogers> changed storing workspaceColor to
> WorkspaceSettings after all.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AndBible/and-bible/pull/2007#issuecomment-1326449483>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKGTJT5U7PJPP5WKH66WHDWJ5T2FANCNFSM5K362MFQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tuomas2 commented:
It was motivated by fixing #2384 which was caused by this.

agrogers commented:
Interesting. I have been using my code since I published it and didn't see
this. Well done on fixing it!

___________________________
*www.TenOutOfTen.org* <https://www.TenOutOfTen.org>
***@***.***
(+61) 0477 174 193 (Australia)
(+95) 09 250018669 (Myanmar)



On Fri, 25 Nov 2022 at 14:15, Tuomas Airaksinen ***@***.***>
wrote:

> It was motivated by fixing #2384
> <https://github.com/AndBible/and-bible/issues/2384> which was caused by
> this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AndBible/and-bible/pull/2007#issuecomment-1327118131>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKGTJQ6E6JZIG2TPIWX5XLWKBU73ANCNFSM5K362MFQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
